ntbs-ui-tests/Hooks/DriverSetup.cs

```diff
@@ -73,6 +73,15 @@ private async Task RemoveImportedNotificationAfter()
 
         private async Task CleanUpMigratedNotification()
         {
+            using (var connection = new SqlConnection(settings.EnvironmentConfig.SpecimenConnectionString))
+            {
+                connection.Open();
+                var deleteSpecimen1 = "exec uspUnmatchSpecimen 'M.7148378', '189045'";
+                var deleteSpecimen2 = "exec uspUnmatchSpecimen 'M.9420326', '189045'";

```

Comment:
This needs to be for the NTBS ID for the newly created notification, rather than the legacy ID here. Sorry that I didn't make this clear enough earlier!

To explain what's going on under the hood: there is a specimen matching table `EtsSpecimenMatch` that contains a bunch of pairings between ETS notification IDs and `ReferenceLabNumber`s (basically specimen IDs), that come from ETS. During migration, we want to migrate these matches over, so the specimens matched to the old ETS notification are also matched to the new NTBS one. To do this we call `uspMatchSpecimen`, which inserts a record into the `NotificationSpecimenMatch` table, which is for NTBS matches, linking the new NTBS ID with the specimen reference lab number.

This stored proc will throw an error if it sees you matching a specimen to an already matched notification. Therefore, you need to remove it from the `NotificationSpecimenMatch` table, so the next time you run the test it can insert a new record with the same specimen reference lab number but a new NTBS ID, which is what you essentially are doing with `uspUnmatchSpecimen`

Reply:
As an aside, this `uspUnmatchSpecimen` SP will return normally, but with an error return result (`-1`) when you run this at the start, when the specimen is currently unmatched. However, this won't cause you any issues, so you can quite happily run this code before and after the test

Reply:
Hm, I think this approach may not work then. There's no way to know, before the test, which NTBS ID correlates to this notification. We could just hope that it hasn't been imported (which it shouldn't have been) and then set this up to correctly remove it after. Would it be possible to go directly to `NotificationSpecimenMatch` and remove the entries manually, or is there other factors that would make that inadvisable?

Reply:
I had assumed you had access to NTBS services when I suggested the method in `SpecimenService`. It is a v good point that you won't know the notification ID at the start of the test because you haven't created it yet!

Yeah I can't foresee any issues with editing the table directly, that's pretty much what the stored procs are doing anyway. `DELETE FROM NotificationSpecimenMatch WHERE ReferenceLaboratoryNumber IN ('M.7148378', 'M.9420326')` is probably your best bet

Reply:
I've changed it to that. Looks like it's meant the specimens are actually matching now on import too - which is good!

Reply:
@LeoCie What you've got here looks fine, but I just wanted to add that you should be able to use the list of `AddedNotificationIds` in the `testContext` when you are trying to clear up things like this.